File: desktop/pes/clock.py
"""Injectable clock so the engine and sync are testable with simulated time.

All engine time is whole epoch seconds (UTC), matching the core.
"""

from __future__ import annotations

import time


class SystemClock:
    def now(self) -> int:
        return int(time.time())


class FakeClock:
    """Settable clock for scenario tests."""

    def __init__(self, epoch: int):
        self._now = epoch

    def now(self) -> int:
        return self._now

    def set(self, epoch: int) -> None:
        self._now = epoch

    def advance(self, seconds: int) -> None:
        self._now += seconds

File: android/runtime/src/main/kotlin/pes/Clock.kt
/** Injectable clock (mirrors `pes/clock.py`): epoch seconds, UTC. */
package pes

interface Clock {
    fun now(): Long
}

class SystemClock : Clock {
    override fun now(): Long = System.currentTimeMillis() / 1000
}

/** Manually advanced clock for scenario tests. */
class FakeClock(var epoch: Long) : Clock {
    override fun now(): Long = epoch

    fun advance(seconds: Long) {
        epoch += seconds
    }

    fun set(newEpoch: Long) {
        epoch = newEpoch
    }
}
